cpp/src/arrow/chunked_array.h

```diff
@@ -41,6 +43,86 @@ template <typename T, typename V>
 class ChunkedArrayIterator;
 }  // namespace stl
 
+class ARROW_EXPORT DeviceAllocationTypeSet {
```

Comment:
Open to suggestion on where to put this class.

Reply:
Would this be something we could add to `Table`, too? Might be worth putting in `device.h` if so.

Reply:
I didn't put in `device.h` because `chunked_array.h` (included a lot!) needs this and `device.h` is a heavy header. So I ended up creating a new `.h` for the type set.